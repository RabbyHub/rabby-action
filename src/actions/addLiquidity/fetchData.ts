import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
