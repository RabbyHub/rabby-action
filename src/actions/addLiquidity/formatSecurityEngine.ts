import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
