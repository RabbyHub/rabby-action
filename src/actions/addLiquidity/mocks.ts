export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
