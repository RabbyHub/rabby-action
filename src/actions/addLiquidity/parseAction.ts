import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
