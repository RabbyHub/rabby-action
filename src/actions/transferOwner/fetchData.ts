import PQueue from 'p-queue';
import {
  FetchActionRequiredData,
  TransferOwnerRequireData,
  ReceiverData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchReceiverHasTransfer } from '../../fetches/fetchHasTransfer';

export const fetchDataTransferOwner: FetchActionRequiredData = async (
  options
) => {
  if (!options.actionData.transferOwner) {
    return {};
  }
  const queue = new PQueue();
  const {
    actionData,
    walletProvider,
    apiProvider,
    chainId,
    sender,
    extraActionDataState,
  } = options;
  const addr = actionData.transferOwner?.to;
  const result: TransferOwnerRequireData = {
    receiver: null,
    extraState: {
      receiverHasTransfer: () => undefined,
    },
  };
  if (addr) {
    const chain = walletProvider.findChain({
      serverId: chainId,
    });
    const receiverData: ReceiverData = {
      address: addr,
      chain: chain!,
      eoa: null,
      cex: null,
      contract: null,
      usd_value: 0,
      hasTransfer: null,
      isTokenContract: false,
      name: null,
      onTransferWhitelist: false,
    };
    fetchReceiverHasTransfer({
      apiProvider,
      chainId,
      sender,
      spender: addr,
      queue,
      extraActionDataState,
      receiverData,
      result,
    });
    queue.add(async () => {
      const { desc } = await apiProvider.addrDesc(addr);
      if (desc.cex?.id) {
        receiverData.cex = {
          id: desc.cex.id,
          logo: desc.cex.logo_url,
          name: desc.cex.name,
          bornAt: desc.born_at,
          isDeposit: desc.cex.is_deposit,
        };
      }
      if (desc.contract && Object.keys(desc.contract).length > 0) {
        receiverData.contract = desc.contract;
      }
      if (!receiverData.cex && !receiverData.contract) {
        receiverData.eoa = {
          id: addr,
          bornAt: desc.born_at,
        };
      }
      receiverData.usd_value = desc.usd_value;
      receiverData.name = desc.name;
    });
    queue.add(async () => {
      if (addr) {
        const { is_token } = await apiProvider.isTokenContract(chainId, addr);
        receiverData.isTokenContract = is_token;
      }

      if (walletProvider.ALIAS_ADDRESS[addr.toLowerCase()]) {
        receiverData.name = walletProvider.ALIAS_ADDRESS[addr.toLowerCase()];
      }

      const whitelist = await walletProvider.getWhitelist();
      receiverData.onTransferWhitelist = whitelist.includes(addr.toLowerCase());
    });
    await waitQueueFinished(queue);
    result.receiver = receiverData;
  }
  return result;
};
