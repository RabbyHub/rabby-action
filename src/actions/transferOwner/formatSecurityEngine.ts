import { TransferOwnerRequireData } from '../../types';
import { FormatSecurityEngineContext } from '../../types';

export const formatSecurityEngineTransferOwner: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId, provider } = options;

    if (!actionData.transferOwner || !chainId) {
      return {};
    }
    const data = requireData as TransferOwnerRequireData;
    return {
      transferOwner: {
        receiverInWhitelist: !!data.receiver?.onTransferWhitelist,
      },
    };
  };
