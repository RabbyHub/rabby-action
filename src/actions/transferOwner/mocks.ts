export const parseTxData = {
  action: {
    type: 'transfer_ownership',
    data: {
      description: 'Transfer proxy ownership.',
      from_addr: '0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4',
      to_addr: '0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4',
    },
  },
  log_id: 6458791,
} as any;

export const txData = {
  from: '0x0000db5c8b030ae20308ac975898e09741e70000',
  to: '0x2129f8a9b6c3092a600da82ce859b7a9a69983e4',
  data: '0x13af40350000000000000000000000005d4b2a02c59197eb2cae95a6df9fe27af60459d4',
  value: '0x0',
  chainId: 1,
  nonce: '0x0',
};
