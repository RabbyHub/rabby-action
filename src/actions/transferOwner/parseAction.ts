import { TransferOwnerAction } from '@rabby-wallet/rabby-api/dist/types';
import { ParseAction } from '../../types';

export const parseActionTransferOwner: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'transfer_ownership') {
    return {};
  }

  const { description, from_addr, to_addr } = data.data as TransferOwnerAction;

  return {
    transferOwner: {
      description,
      from: from_addr,
      to: to_addr,
    },
  };
};
