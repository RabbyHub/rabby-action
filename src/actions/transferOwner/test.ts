import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  ORIGIN,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { fetchDataTransferOwner } from './fetchData';
import { formatSecurityEngineTransferOwner } from './formatSecurityEngine';
import { parseTxData, txData } from './mocks';
import { parseActionTransferOwner } from './parseAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';

/**
 * https://brain.debank.com/action-group/tx/0x13af4035
 * id: 83085816
 * chainId: eth
 * contract_id: 0x2129f8a9b6c3092a600da82ce859b7a9a69983e4
 */
test.each([
  [
    parseTypedDataAction(parseActionTransferOwner),
    fetchDataTransferOwner,
    formatSecurityEngineTransferOwner,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('TransferOwner', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'typed_data',
    data: parseTxData['action'],
    typedData: txData,
    sender: SENDER,
  });
  expect(actionData).toMatchSnapshot('parseActionTransferOwner');

  const requireData = await _fetchData({
    type: 'typed_data',
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    apiProvider,
    actionData,
  });

  expect(requireData).toMatchSnapshot('fetchDataTransferOwner');

  const ctx = await _format({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineTransferOwner');
});
