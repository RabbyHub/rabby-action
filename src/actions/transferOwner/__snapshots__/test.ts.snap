// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferOwner: fetchDataTransferOwner 1`] = `
{
  "extraState": {
    "receiverHasTransfer": [Function],
  },
  "receiver": {
    "address": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    "cex": null,
    "chain": {
      "eip": {
        "1559": true,
      },
      "enum": "ETH",
      "hex": "0x1",
      "id": 1,
      "logo": {},
      "name": "Ethereum",
      "nativeTokenAddress": "eth",
      "nativeTokenDecimals": 18,
      "nativeTokenLogo": "https://static.debank.com/image/token/logo_url/eth/935ae4e4d1d12d59a99717a24f2540b5.png",
      "nativeTokenSymbol": "ETH",
      "network": "1",
      "scanLink": "https://etherscan.io/tx/_s_",
      "serverId": "eth",
      "whiteLogo": {},
    },
    "contract": null,
    "eoa": {
      "bornAt": 1564655163,
      "id": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    },
    "hasTransfer": true,
    "isTokenContract": true,
    "name": null,
    "onTransferWhitelist": false,
    "usd_value": 4418.154405584272,
  },
}
`;

exports[`TransferOwner: fetchDataTransferOwner 2`] = `
{
  "extraState": {
    "receiverHasTransfer": [Function],
  },
  "receiver": {
    "address": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    "cex": null,
    "chain": {
      "eip": {
        "1559": true,
      },
      "enum": "ETH",
      "hex": "0x1",
      "id": 1,
      "logo": {},
      "name": "Ethereum",
      "nativeTokenAddress": "eth",
      "nativeTokenDecimals": 18,
      "nativeTokenLogo": "https://static.debank.com/image/token/logo_url/eth/935ae4e4d1d12d59a99717a24f2540b5.png",
      "nativeTokenSymbol": "ETH",
      "network": "1",
      "scanLink": "https://etherscan.io/tx/_s_",
      "serverId": "eth",
      "whiteLogo": {},
    },
    "contract": null,
    "eoa": {
      "bornAt": 1564655163,
      "id": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    },
    "hasTransfer": true,
    "isTokenContract": true,
    "name": null,
    "onTransferWhitelist": false,
    "usd_value": 4418.154405584272,
  },
}
`;

exports[`TransferOwner: formatSecurityEngineTransferOwner 1`] = `
{
  "transferOwner": {
    "receiverInWhitelist": false,
  },
}
`;

exports[`TransferOwner: formatSecurityEngineTransferOwner 2`] = `
{
  "transferOwner": {
    "receiverInWhitelist": false,
  },
}
`;

exports[`TransferOwner: parseActionTransferOwner 1`] = `
{
  "actionType": "transfer_ownership",
  "brand": undefined,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "transferOwner": {
    "description": "Transfer proxy ownership.",
    "from": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    "to": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
  },
}
`;

exports[`TransferOwner: parseActionTransferOwner 2`] = `
{
  "actionType": "transfer_ownership",
  "brand": undefined,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "transferOwner": {
    "description": "Transfer proxy ownership.",
    "from": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
    "to": "0x5d4b2a02c59197eb2cae95a6df9fe27af60459d4",
  },
}
`;
